connectors/sources/slack.py

```diff
@@ -0,0 +1,393 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+
+import re
+import time
+from contextlib import asynccontextmanager
+from datetime import datetime, timedelta
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource
+from connectors.utils import CancellableSleeps, dict_slice, retryable
+
+BASE_URL = "https://slack.com/api"
+
+CURSOR = "cursor"
+RESPONSE_METADATA = "response_metadata"
+NEXT_CURSOR = "next_cursor"
+DEFAULT_RETRY_SECONDS = 3
+DEFAULT_RETRY_SLEEP = 30
+PAGE_SIZE = 200
+USER_ID_PATTERN = re.compile(r"<@([A-Z0-9]+)>")
+
+# TODO list
+# Nice to haves:
+# - expand links
+#   - links:read scope needed?
+# - configure which channels to sync via deny list
+# - write an ftest
+
+
+class ThrottledError(Exception):
+    """Internal exception class to indicate that request was throttled by the API"""
+
+    pass
+
+
+class SlackClient:
+    def __init__(self, configuration):
+        self.token = configuration["token"]
+        self._http_session = aiohttp.ClientSession(
+            headers=self._headers(),
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._logger = logger
+        self._sleeps = CancellableSleeps()
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    def ping(self):
+        return self.test_auth()
```

Comment:
Wondering why defining another `test_auth` method.

Reply:
Because the endpoint is "auth.test", so the client should have a "test auth" method. It's just an implementation detail that ping (a framework convention of ours) uses the test_auth method.